File: Algorithm.CSharp/FutureStopMarketOrderOnExtendedHoursRegressionAlgorithm.cs
/*
 * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
 * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

using System;
using System.Linq;
using QuantConnect.Data;
using QuantConnect.Orders;
using QuantConnect.Securities;
using QuantConnect.Interfaces;
using QuantConnect.Data.Market;
using System.Collections.Generic;
using QuantConnect.Securities.Future;

namespace QuantConnect.Algorithm.CSharp
{
    /// <summary>
    /// Continuous Futures Regression algorithm.
    /// Asserting the behavior of stop market order <see cref="StopMarketOrder"/> in extended market hours
    /// <seealso cref="Data.UniverseSelection.UniverseSettings.ExtendedMarketHours"/>
    /// </summary>
    public class FutureStopMarketOrderOnExtendedHoursRegressionAlgorithm : QCAlgorithm, IRegressionAlgorithmDefinition
    {
        private OrderTicket _ticket;
        private Future _SP500EMini;
        public override void Initialize()
        {
            SetStartDate(2013, 10, 6);
            SetEndDate(2013, 10, 12);

            _SP500EMini = AddFuture(Futures.Indices.SP500EMini, Resolution.Minute, extendedMarketHours: true);

            Schedule.On(DateRules.EveryDay(), TimeRules.At(19, 0), () =>
            {
                // Don't place orders at the end of the last date, the market-on-stop order won't have time to fill
                if (Time.Date == EndDate.Date.AddDays(-1))
                {
                    return;
                }

                MarketOrder(_SP500EMini.Mapped, 1);
                _ticket = StopMarketOrder(_SP500EMini.Mapped, -1, _SP500EMini.Price * 1.1m);
            });
        }

        /// <summary>
        /// Data Event Handler: receiving all subscription data in a single event
        /// </summary>
        /// <param name="slice">The current slice of data keyed by symbol string</param>
        public override void OnData(Slice slice)
        {
            if (_ticket == null || _ticket.Status != OrderStatus.Submitted)
            {
                return;
            }

            var stopPrice = _ticket.Get(OrderField.StopPrice);
            var bar = Securities[_ticket.Symbol].Cache.GetData<TradeBar>();
        }

        /// <summary>
        /// Order fill event handler. On an order fill update the resulting information is passed to this method.
        /// </summary>
        /// <param name="orderEvent">Order event details containing details of the events</param>
        public override void OnOrderEvent(OrderEvent orderEvent)
        {
            if (orderEvent == null)
            {
                return;
            }

            if (Transactions.GetOrderById(orderEvent.OrderId).Type != OrderType.StopMarket)
            {
                return;
            }

            if (orderEvent.Status == OrderStatus.Filled)
            {
                var time = MarketHoursDatabase.GetExchangeHours(_SP500EMini.SubscriptionDataConfig);

                if (!time.IsOpen(orderEvent.UtcTime, _SP500EMini.IsExtendedMarketHours))
                {
                    throw new RegressionTestException($"The Exchange hours was closed, verify 'extendedMarketHours' flag in {nameof(Initialize)} when added new security(ies).");
                }
            }
        }

        public override void OnEndOfAlgorithm()
        {
            var stopMarketOrders = Transactions.GetOrders(x => x is StopMarketOrder);

            if (stopMarketOrders.Any(x => x.Status != OrderStatus.Filled))
            {
                throw new RegressionTestException("The Algorithms was not handled any StopMarketOrders");
            }
        }

        /// <summary>
        /// This is used by the regression test system to indicate if the open source Lean repository has the required data to run this algorithm.
        /// </summary>
        public bool CanRunLocally { get; } = true;

        /// <summary>
        /// This is used by the regression test system to indicate which languages this algorithm is written in.
        /// </summary>
        public List<Language> Languages { get; } = new() { Language.CSharp, Language.Python };

        /// <summary>
        /// Data Points count of all time slices of algorithm
        /// </summary>
        public long DataPoints => 41486;

        /// <summary>
        /// Data Points count of the algorithm history
        /// </summary>
        public int AlgorithmHistoryDataPoints => 0;

        /// <summary>
        /// Final status of the algorithm
        /// </summary>
        public AlgorithmStatus AlgorithmStatus => AlgorithmStatus.Completed;

        /// <summary>
        /// This is used by the regression test system to indicate what the expected statistics are from running the algorithm
        /// </summary>
        public Dictionary<string, string> ExpectedStatistics => new Dictionary<string, string>
        {
            {"Total Orders", "10"},
            {"Average Win", "0%"},
            {"Average Loss", "-0.02%"},
            {"Compounding Annual Return", "-6.419%"},
            {"Drawdown", "0.100%"},
            {"Expectancy", "-1"},
            {"Start Equity", "100000"},
            {"End Equity", "99891"},
            {"Net Profit", "-0.109%"},
            {"Sharpe Ratio", "-22.29"},
            {"Sortino Ratio", "-26.651"},
            {"Probabilistic Sharpe Ratio", "0.016%"},
            {"Loss Rate", "100%"},
            {"Win Rate", "0%"},
            {"Profit-Loss Ratio", "0"},
            {"Alpha", "-0.05"},
            {"Beta", "-0.006"},
            {"Annual Standard Deviation", "0.002"},
            {"Annual Variance", "0"},
            {"Information Ratio", "-2.76"},
            {"Tracking Error", "0.215"},
            {"Treynor Ratio", "8.829"},
            {"Total Fees", "$21.50"},
            {"Estimated Strategy Capacity", "$3400000.00"},
            {"Lowest Capacity Asset", "ES VMKLFZIH2MTD"},
            {"Portfolio Turnover", "138.95%"},
            {"OrderListHash", "957191893a3de4975ec14b2a3b2490de"}
        };
    }
}

File: Algorithm.Python/FutureStopMarketOrderOnExtendedHoursRegressionAlgorithm.py
# QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
# Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from AlgorithmImports import *

# <summary>
# This example demonstrates how to create future 'stop_market_order' in extended Market Hours time
# </summary>

class FutureStopMarketOrderOnExtendedHoursRegressionAlgorithm(QCAlgorithm):
    # Keep new created instance of stop_market_order
    _stop_market_ticket = None
    
    # Initialize the Algorithm and Prepare Required Data
    def initialize(self) -> None:
        self.set_start_date(2013, 10, 6)
        self.set_end_date(2013, 10, 12)

        # Add mini SP500 future with extended Market hours flag
        self._sp_500_e_mini = self.add_future(Futures.Indices.SP_500_E_MINI, Resolution.MINUTE, extended_market_hours=True)

        # Init new schedule event with params: every_day, 19:00:00 PM, what should to do
        self.schedule.on(self.date_rules.every_day(),self.time_rules.at(19, 0),self.make_market_and_stop_market_order)

    # This method is opened 2 new orders by scheduler
    def make_market_and_stop_market_order(self) -> None:
        # Don't place orders at the end of the last date, the market-on-stop order won't have time to fill
        if self.time.date() == self.end_date.date() - timedelta(1) or  not self._sp_500_e_mini.mapped:
            return

        self.market_order(self._sp_500_e_mini.mapped, 1)
        self._stop_market_ticket = self.stop_market_order(self._sp_500_e_mini.mapped, -1, self._sp_500_e_mini.price * 1.1)

    # New Data Event handler receiving all subscription data in a single event
    def on_data(self, slice: Slice) -> None:
        if (self._stop_market_ticket == None or self._stop_market_ticket.status != OrderStatus.SUBMITTED):
            return None

        self.stop_price = self._stop_market_ticket.get(OrderField.STOP_PRICE)
        self.bar = self.securities[self._stop_market_ticket.symbol].cache.get_data()

    # An order fill update the resulting information is passed to this method.
    def on_order_event(self, order_event: OrderEvent) -> None:
        if self.transactions.get_order_by_id(order_event.order_id).type is not OrderType.STOP_MARKET:
            return None

        if order_event.status == OrderStatus.FILLED:
            # Get Exchange Hours for specific security
            exchange_hours = self.market_hours_database.get_exchange_hours(self._sp_500_e_mini.subscription_data_config)

            # Validate, Exchange is opened explicitly
            if (not exchange_hours.is_open(order_event.utc_time, self._sp_500_e_mini.is_extended_market_hours)):
                raise AssertionError("The Exchange hours was closed, verify 'extended_market_hours' flag in Initialize() when added new security(ies)")

    def on_end_of_algorithm(self) -> None:
        self.stop_market_orders = self.transactions.get_orders(lambda o: o.type is OrderType.STOP_MARKET)

        for o in self.stop_market_orders:
            if o.status != OrderStatus.FILLED:
                raise AssertionError("The Algorithms was not handled any StopMarketOrders")
